OpenRA.Game/Graphics/TerrainSpriteLayer.cs

```diff
@@ -191,32 +218,54 @@ public void Update(MPos uv, Sprite sprite, PaletteReference palette, in float3 p
 				UpdateTint(uv);
 			}
 
-			dirtyRows.Add(uv.V);
+			MarkDirty(uv);
 		}
 
 		public void Draw(Viewport viewport)
 		{
 			var cells = restrictToBounds ? viewport.VisibleCellsInsideBounds : viewport.AllVisibleCells;
 
-			// Only draw the rows that are visible.
-			var firstRow = cells.CandidateMapCoords.TopLeft.V.Clamp(0, map.MapSize.Y);
-			var lastRow = (cells.CandidateMapCoords.BottomRight.V + 1).Clamp(firstRow, map.MapSize.Y);
+			// Only update and draw bins that are visible
+			var tl = cells.CandidateMapCoords.TopLeft;
+			var br = cells.CandidateMapCoords.BottomRight;
+			var minCol = tl.U.Clamp(0, map.MapSize.X) / binSize;
+			var maxCol = br.U.Clamp(0, map.MapSize.X) / binSize;
+			var minRow = tl.V.Clamp(0, map.MapSize.Y) / binSize;
+			var maxRow = br.V.Clamp(0, map.MapSize.Y) / binSize;
 
-			Game.Renderer.Flush();
+			Game.Renderer.WorldSpriteRenderer.SetRenderStateForVertexBuffer(vertexBuffer, sheets, BlendMode);
 
 			// Flush any visible changes to the GPU
-			for (var row = firstRow; row <= lastRow; row++)
+			for (var row = minRow; row <= maxRow; row++)
 			{
-				if (!dirtyRows.Remove(row))
-					continue;
+				for (var col = minCol; col <= maxCol;)
+				{
+					var i = row * binCols + col;
+					if (!dirtyBins[i])
+					{
+						col++;
+						continue;
+					}
+
+					// Coalesce adjacent bin updates to reduce SetData calls
+					var updateStart = i * binStride;
+					var updateLength = binStride;
+					dirtyBins[i] = false;
+
+					while (++col <= maxCol && dirtyBins[++i])
+					{
+						updateLength += binStride;
+						dirtyBins[i] = false;
+					}
+
```

Comment:
This would add a lot of complexity for negigible benefit - the cells are organized as rows within each bin, so this would require that we chop the bin up and either reindex it or draw it with multiple sub-draws. Remember that thanks to the visible map bounds we may have bins with empty cells on any of its four edges!

Reply:
Re coalescing across multiple rows: this would only happen if the viewer was zoomed out far enough to have the whole map visible on screen, which is a very rare edge case.